### PR TITLE
[DOC-11277] Document FORMALIZE miscellaneous function

### PR DESCRIPTION
**Pull request by julia-cb**
Jira ticket: [DOC-11277](https://issues.couchbase.com/browse/DOC-11277)

Page affected: [Miscellaneous Utility Functions](https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/metafun.html)

Changes in this PR:

- Added section on FORMALIZE(`statement` [ `,query_context` ]) to miscellaneous functions page.

(Please ignore the Eventing page that was deleted!! I ended up pushing an Eventing change by accident and couldn't figure out how to remove it from the PR without deleting the file and re-pushing the changes. Git is confusing sometimes 😵‍💫)

### CODE REVIEW

File: modules/n1ql/pages/n1ql-language-reference/metafun.adoc [hunk 1]
```diff
@@ -416,6 +416,46 @@ For any item not listed above, the return value is `null`.
...
 === Example
 
+[[formalize,FORMALIZE()]]
+== FORMALIZE(`statement` [ `,query_context` ])
+
+=== Description
+
+Specifies all references from a query.
+
+Can be spelled as FORMALIZE or FORMALISE.
```
Comment (lamagnifica) [inline modules/n1ql/pages/n1ql-language-reference/metafun.adoc]: fragments ok?

File: modules/n1ql/pages/n1ql-language-reference/metafun.adoc [hunk 2]
```diff
@@ -416,6 +416,46 @@ For any item not listed above, the return value is `null`.
...
+Returns a query with all references fully specified.
+
+=== Example
+
+[[formalize-ex,FORMALIZE() Example]]
+====
+.Query
+[source,sqlpp]
+----
+SELECT FORMALIZE("select * from c1","default: example")
```
Comment (lamagnifica) [inline modules/n1ql/pages/n1ql-language-reference/metafun.adoc]: is there a way to do use the travel-sample.inventory data to show this? without knowing the background of what needs to be formalized, seeing the result doesn't really help

Comment (julia-cb) [reply modules/n1ql/pages/n1ql-language-reference/metafun.adoc]: Good question - I'll update the example to use the travel-sample data!

File: modules/n1ql/pages/n1ql-language-reference/metafun.adoc [hunk 3]
```diff
@@ -368,6 +368,70 @@ The return value is a flattened list of array elements for use in an array index
 
 For examples, refer to xref:n1ql:n1ql-language-reference/indexing-arrays.adoc#examples[Array Indexing Examples].
 
+[[formalize,FORMALIZE()]]
+== FORMALIZE(`statement` [ `,query_context` ])
+
```
Comment (simon-dew) [inline modules/n1ql/pages/n1ql-language-reference/metafun.adoc:373]: Standard boilerplate text for new feature

```asciidoc

ifeval::['{page-component-version}' == '7.6']
_(Introduced in Couchbase Server 7.6)_
endif::[]

```

File: modules/n1ql/pages/n1ql-language-reference/metafun.adoc [hunk 4]
```diff
@@ -368,6 +368,70 @@ The return value is a flattened list of array elements for use in an array index
...
+=== Description
+
+Specifies all references from a query.
```
Comment (simon-dew) [inline modules/n1ql/pages/n1ql-language-reference/metafun.adoc]: Looking for slightly more descriptive wording here. Something like

```suggestion
Fully expands all references within a query, using the specified query context.
```

File: modules/n1ql/pages/n1ql-language-reference/metafun.adoc [hunk 5]
```diff
@@ -368,6 +368,70 @@ The return value is a flattened list of array elements for use in an array index
...
+
+Can be spelled as FORMALIZE or FORMALISE.
```
Comment (simon-dew) [inline modules/n1ql/pages/n1ql-language-reference/metafun.adoc]:
```suggestion
This function has a synonym FORMALISE().
```

Check UNNEST_POS and UNNEST_POSITION in this document to see how synonyms are handled.

File: modules/n1ql/pages/n1ql-language-reference/metafun.adoc [hunk 6]
```diff
@@ -368,6 +368,70 @@ The return value is a flattened list of array elements for use in an array index
...
+A string query context value for the function to use when formalizing.
+
+=== Return Value
+
+Returns a query with all references fully specified.
+
+=== Examples
+
+[[formalize-ex,FORMALIZE() Example]]
+====
```
Comment (simon-dew) [inline modules/n1ql/pages/n1ql-language-reference/metafun.adoc]: As there's more than one example for FORMALIZE(), let's give the example a title

```suggestion
[[formalize-ex1,FORMALIZE() Example 1]]
.Formalize a query
====
```

File: modules/n1ql/pages/n1ql-language-reference/metafun.adoc [hunk 7]
```diff
@@ -368,6 +368,70 @@ The return value is a flattened list of array elements for use in an array index
...
+----
+[
+  {
+    "$1": "select self.* from `default`:`travel-sample`.`inventory`.`landmark` where ((`landmark`.`country`) = \"United Kingdom\")"
+  }
+]
+----
+====
+
+====
```
Comment (simon-dew) [inline modules/n1ql/pages/n1ql-language-reference/metafun.adoc:418]: And let's give this example an ID and a title too

```suggestion
[[formalize-ex2,FORMALIZE() Example 2]]
.Formalize recently completed requests
====
```

File: modules/n1ql/pages/n1ql-language-reference/metafun.adoc [hunk 8]
```diff
@@ -368,6 +368,70 @@ The return value is a flattened list of array elements for use in an array index
...
+.Results
+[source,json]
+----
+[
+  {
+    "statement": "advise SELECT *\nFROM route\nJOIN LATERAL airline\nON route.airlineid = META(airline).id\nWHERE route.sourceairport = \"SFO\";",
+    "queryContext": "default:travel-sample.inventory",
+    "formalized": "advise SELECT *\nFROM route\nJOIN LATERAL airline\nON route.airlineid = META(airline).id\nWHERE route.sourceairport = \"SFO\";"
+  },
+  ...
```
Comment (simon-dew) [inline modules/n1ql/pages/n1ql-language-reference/metafun.adoc]: I know it's not standard JSON but I usually add a JSON-C comment marker for ellipses within a JSON file

```suggestion
  // ...
```

File: modules/n1ql/pages/n1ql-language-reference/metafun.adoc [hunk 9]
```diff
@@ -368,6 +368,70 @@ The return value is a flattened list of array elements for use in an array index
...
+====
+
+====
+.Query
+[source,sqlpp]
+----
+SELECT statement
+      ,NVL(queryContext,"") AS queryContext
+      ,formalize(statement, queryContext) AS formalized 
+FROM system:completed_requests;
```
Comment (simon-dew) [inline modules/n1ql/pages/n1ql-language-reference/metafun.adoc]: Looks kind of odd having the commas at the start of the line. Reads better if they're at the end of the previous line.

```suggestion
SELECT statement,
       NVL(queryContext,"") AS queryContext,
       formalize(statement, queryContext) AS formalized 
FROM system:completed_requests;
```

**Review by simon-dew — CHANGES_REQUESTED** [review]
Brilliant. Just a few minor changes.

**Comment by simon-dew** [comment]
When we're finished with peer review, please add @dhaggart for SME review, as the person who raised the original issue.

**Comment by julia-cb** [comment]
@dhaggart Hi Donald, I added the new FORMALIZE function to the docs. Please take a quick look and review when you have a sec! Thanks :)

File: modules/n1ql/pages/n1ql-language-reference/metafun.adoc [hunk 10]
```diff
@@ -368,6 +368,77 @@ The return value is a flattened list of array elements for use in an array index
...
+       formalize(statement, queryContext) AS formalized 
+FROM system:completed_requests;
+
+----
+.Results
+[source,json]
+----
+[
+  {
+    "statement": "advise SELECT *\nFROM route\nJOIN LATERAL airline\nON route.airlineid = META(airline).id\nWHERE route.sourceairport = \"SFO\";",
```
Comment (dhaggart) [inline modules/n1ql/pages/n1ql-language-reference/metafun.adoc]: I'd avoid using an "advise" statement in the example.  I'd also pick one where we're actually seeing it be transformed.

**Review by dhaggart — CHANGES_REQUESTED** [review]

File: modules/n1ql/pages/n1ql-language-reference/metafun.adoc (hunk 10 above)
Comment (julia-cb) [reply modules/n1ql/pages/n1ql-language-reference/metafun.adoc]: Got it - I updated the example with the results you sent me over slack. Let me know if that's better!

**Review by dhaggart — APPROVED** [review]

**Pull request merged**